specification/taf-spec.md

```diff
@@ -0,0 +1,915 @@
+target files named after these repositories. The first step is to extract lists of commits per branch
+for each target repository. These target files are JSON files containing data in this format:
+
+```markdown
+{
+  "branch": "main",
+  "commit": "e3f4g5h6"
+}
+```
+
+- Fetch new commits (without pulling, which includes merging) or clone target repositories without
+checking out files for additional safety. This step avoids merging changes into local branches and
+prevents immediate exposure to the files in the repository. The URLs of the target repositories are
+determined based on the `repositories.json` and `mirrors.json` files, as explained earlier.
+- Compare commits extracted from the authentication repository with the fetched ones in a breadth-first
+manner. This involves checking each target repository against the target files contained in a given
+authentication repository's commit, then moving to the next commit. The process stops if there is a
+mismatch.
+- If a repository is allowed to contain unauthenticated commits, an error is not reported as long as
+all commits listed in the authentication repository are found in that target repository in the correct
+order, even if there are additional unlisted commits.
+- After the validation process is completed, whether it ends successfully or with an error, the next
+action is to merge the last successfully validated commit into each respective repository.
+
+After successfully validating the target repositories, the final step is to merge the last validated
+authentication commit into the local copy of the authentication repository. Once this is done,
+record this commit SHA in the `last_validated_commit` file. This action ensures that the local
+authentication repository is up-to-date with the latest validated state and that this state
+is accurately recorded for future reference.
+
+An important aspect of TAF's update process is the determination of destination paths for target
+repositories on the filesystem. When initiating the updater, an optional input parameter can be
+provided that specifies the root location for all repositories. For instance, if this root location
+is set to /dir1/dir2, the path for a target repository will be formed by appending its full name to
+this path. This would result in a path like /dir1/dir2/namespace/name1 for the target repository.
+
+In cases where this root path is not explicitly provided, TAF defaults to determining it based on the
+path of the authentication repository. It assumes that the full name of the authentication repository
+is included in its path, and the root for the target repositories is located two directories up from
+the authentication repository's location."
+
+Suppose the path of the authentication repository is `/dir1/dir2/namespace/auth_repo`.
+Considering that the authentication repository's full name (`namespace/auth_repo`) is included
+in its path, TAF calculates the root location for target repositories as `/dir1/dir2`.
```

Comment:
These kind of look like implementation details for finding paths on local disk. Are they necessary in the spec? Or should they maybe go into a more detailed document that lists the CLI behavior? Maybe [here ](https://github.com/openlawlibrary/taf/blob/master/docs/repo-creation-and-update-util.md) or [here](https://github.com/openlawlibrary/taf/blob/master/docs/setting-up-new-library.md)?